lib/util/historical-buffer-manager.ts

```diff
@@ -0,0 +1,95 @@
+/// <reference path='../../typings/tsd.d.ts' />
+
+import assert = require('assert');
+import interfaces = require('../interface');
+
+// Internal class that holds the buffer array, a counter for overflow items and maxBufferLength
+class Buffer<T> {
+    data: T[];
+    overflow: number;
+    maxBufferLength: number;
+
+    get getObject(): interfaces.IBufferedData<T> {
+        return { buffer: this.data.slice(),
+                 overflow: this.overflow };
+    }
+
+    constructor(max: number) {
+        this.data = [];
+        this.overflow = 0;
+        this.maxBufferLength = max;
+    }
+}
+
+export = HistoricalBufferManager;
+
+// This class creates a buffer manager which manage N buffers(N is taken from constructor parameter)
+class HistoricalBufferManager<T> {
+
+    // PRIVATE VARIABLES
+    private _buffers: Buffer<T>[];
+    private _currentIndex: number;
+    private _historyDepth: number;
+    private _defaultBuffLength: number;
+
+    // maxHistoryDepth: Maximum numbers of preserved buffer history(default value: 1)
+    // maxBufferLength: Maximum numbers of data item per level of buffer(default value: 100)
+    constructor(maxHistoryDepth: number = 1, maxBufferLength: number = 100) {
+        assert(maxHistoryDepth >= 1, 'Invalid maxHistoryDepth parameter ' + maxHistoryDepth);
+        assert(maxBufferLength >= 1, 'Invalid maxBufferLength parameter ' + maxBufferLength);
+        this._historyDepth = maxHistoryDepth;
+        this._defaultBuffLength = maxBufferLength;
+
+        // Initialize buffers
+        this._buffers = [];
+        this._currentIndex = 0;
+        for (var i = 0; i < this._historyDepth; ++i) {
+            this._buffers.push(new Buffer<T>(this._defaultBuffLength));
+        }
+    }
+
+    // Push one data item to the current buffer
+    pushValue(data: T): void {
+        var currBuffer: Buffer<T> = this._buffers[this._currentIndex];
+        if (currBuffer.data.length === currBuffer.maxBufferLength) {
+            currBuffer.data.splice(0, 1);
+            ++currBuffer.overflow;
```

Comment:
I think it is common to have this be `currBuffer.overflow++` and it is more clear what is being incremented.